src/python/WMComponent/RucioInjector/RucioInjectorPoller.py

```diff
@@ -370,15 +370,84 @@ def deleteBlocks(self):
         """
         _deleteBlocks_
         Find deletable blocks, then decide if to delete based on:
-        Is there an active subscription for dataset or block ?
-          If yes => set deleted=2
-          If no => next check
         Has transfer to all destinations finished ?
-          If yes => request block deletion, approve request, set deleted=1
+          If yes => Delete rules associated with the block, set deleted=1
           If no => do nothing (check again next cycle)
         """
-        # FIXME: figure out the proper logic for rule block deletion
-        logging.info("Starting deleteBlocks methods --> IMPLEMENT-ME!!!")
+        logging.info("Preparing to create block rules into Rucio...")
+
+        # Get list of blocks that can be deleted
+        blockDict = self.findDeletableBlocks.execute(transaction=False)
+
+        if not blockDict:
+            logging.info("No candidate blocks found for rule deletion")
+            return
+
+        logging.info("Found %d candidate blocks for rule deletion", len(blockDict))
+
+        blocksToDelete = []
+        containerDict = {}
+        # Populate containerDict, assigning each block to its correspondant container
+        for blockName in blockDict:
+            container = blockDict[blockName]['dataset']
+            # If the container is not in the dictionary, create a new entry for it
+            if container not in containerDict:
+                # Set of sites to which the container needs to be transferred
+                sites = blockDict[blockName]['sites']
+                containerDict[container] = {'blocks': [], 'rse': sites}
+            containerDict[container]['blocks'].append(blockName)
+
+        for contName in containerDict:
+            cont = containerDict[contName]
+
+            # Checks if the container is not requested in any sites.
+            # This should never be triggered, but better safe than sorry
+            if not cont['rse']:
+                logging.warning("No rules for container: %s. Its blocks won't be deleted.", contName)
+                continue
+
+            try:
+                # Get RSE in which each block is available
+                availableRSEs = self.rucio.getReplicaInfoForBlocks(block=cont['blocks'])
+            except Exception as exc:
+                msg = "Failed to get replica info for blocks in container: %s.\n" % contName
+                msg += "Will retry again in the next cycle. Error: %s" % str(exc)
+                logging.error(msg)
+                continue
+
+            for blockRSEs in availableRSEs:
+                # If block is available at every RSE its container needs to be transferred, the block can be deleted
+                blockSites = set(blockRSEs['replica'])
+                if cont['rse'].issubset(blockSites) or len(blockSites & cont['rse']) >= 2:
```

Comment:
This logic isn't very clear to me. Trying to expand it in plain english:
* if block is available in a subset of the container locations, then delete it (even if block is available in a single location - do we want this behaviour?)
* or, if block location and container location have at least 2 common RSEs, delete it (does it cover the parked data, which requires a single archival copy?)
Is that correct?

Reply:
The conditions are the following:
-The first condition is: The container locations is a subset of the RSEs in which the block is available. If this condition is met, it means the block has been transferred to every site to which a container level rule has been created, therefore, it can be removed. This is the way PhEDExInjector checks if the block can be deleted. 
- The second condition checks if the block is available in at least 2 of the sites to which its container was subscribed. If this condition is met, the block is in enough sites so it can be deleted.

Taking into account that I added the second condition, the first condition will only make a difference in the case of blocks belonging to containers that are only subscribed to 1 site. This can cover a few of the parked data scenarios, but it's probably best to remove the first condition and create a check to deal specifically with parked data. I'm not sure how to do that in a not-hardcoded-way

Reply:
After discussion with @drkovalskyi , It is clear that we only need the first condition. This method will assume that the container rules meet the Tier0 requirements, and each block will be deleted if and only if it has been transferred to every requested destination.

Reply:
Yes, it looks good. And L405 protects you against containers without any data placement!